execution/block-partitioner/src/v2/union_find.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright © Aptos Foundation
+
+use std::cmp::Ordering;
+
+pub struct UnionFind {
+    parent_of: Vec<usize>,
+    depth_of: Vec<usize>,
+}
+
+impl UnionFind {
+    pub fn new(num_participants: usize) -> Self {
+        Self {
+            parent_of: (0..num_participants).collect(),
+            depth_of: vec![0; num_participants],
+        }
+    }
+
+    pub fn find(&mut self, a: usize) -> usize {
+        let pa = self.parent_of[a];
+        if pa == a {
+            return a;
+        }
+        let ppa = self.find(pa);
```

Comment:
This recursive path compression in UF, uses extra memory (stack memory of recursion).
It could be quicker to do path compression by just using while().. Check this out
https://algs4.cs.princeton.edu/15uf/WeightedQuickUnionPathCompressionUF.java.html